src/app/shared/spaces.service.ts

```diff
@@ -1,72 +1,92 @@
-        .map(val => val.space),
-      this.broadcaster.on<Space>('spaceUpdated'));
-    // Create the recent context list
+      this.contexts.current.map(val => val.space),
+      this.broadcaster.on<Space>('spaceUpdated')
+    );
+  }
+
+  private initRecent(): void {
+    // load existing recent spaces from profile.store.recentSpaces
+    this.loadRecent().subscribe((recentSpaces: Space[]) => {
+      this._recent.next(recentSpaces);
+    });
 
-    this._recent = Observable.merge(
-      this.broadcaster.on<Space>('spaceChanged'),
-      this.addRecent)
-      // Map from the context being added to an array of recent contexts
-      // The scan operator allows us to access the list of recent contexts and add ours
-      .scan((recent, s) => {
-        // First, check if this context is already in the list
-        // If it is, remove it, so we don't get duplicates
-        for (let i = recent.length - 1; i >= 0; i--) {
-          if (recent[i].id === s.id) {
-            recent.splice(i, 1);
+    // update _recent with newly seen space when spaceChanged is emitted
+    this.broadcaster.on<Space>('spaceChanged')
+      .flatMap((changedSpace: Space) => {
+        return this._recent.map((recentSpaces: Space[]) => {
+          let index: number = this.findRecentIndexById(recentSpaces, changedSpace.id);
+          // if changedSpace exists in _recent
+          if (index !== -1) {
+            // move the space to the front of the list if it is already in _recent
+            if (recentSpaces[0].id !== changedSpace.id) {
+              recentSpaces.splice(index, 1);
+              recentSpaces.unshift(changedSpace);
+              this.recentChanged = true;
+            }
+            // otherwise, no operation required; changedSpace is already index 0
+          // if changedSpace is new to the recent spaces
+          } else {
+            recentSpaces.unshift(changedSpace);
+            // trim _recent if required to ensure it is length =< 8
+            if (recentSpaces.length > SpacesService.RECENT_SPACE_LENGTH) {
+              recentSpaces.pop();
+            }
+            this.recentChanged = true;
           }
-        }
-        // Then add this context to the top of the list
-        recent.unshift(s);
-        return recent;
-        // The final value to scan is the initial value, used when the app starts
-      },
-      [])
-      // Finally save the list of recent contexts
-      .do(val => {
-        // Truncate the number of recent contexts to the correct length
-        if (val.length > SpacesService.RECENT_SPACE_LENGTH) {
-          val.splice(
-            SpacesService.RECENT_SPACE_LENGTH,
-            val.length - SpacesService.RECENT_SPACE_LENGTH
-          );
-        }
+          return recentSpaces;
+        });
       })
-      .do(val => {
-        this.saveRecent(val);
+      .subscribe((recentSpaces: Space[]) => {
+        if (this.recentChanged) {
+          this.saveRecent(recentSpaces);
```

Comment:
These should take the first of `_recent` and emit a new item to `_recent` via `_recent.next(item)`, no?

Reply:
Yeah that sounds right I think, so it takes the first when subscribing [0] and if need be it emits the new spaces [1]

[0] https://github.com/fabric8-ui/fabric8-ui/pull/3211/files#diff-103682c5eace85417d403c7c7643dcb5R42
[1] https://github.com/fabric8-ui/fabric8-ui/pull/3211/files#diff-103682c5eace85417d403c7c7643dcb5R69

Reply:
Hum, the branch I checked out this morning didn't have these... Anyways, yes it looks good.

What's the purpose of the `recentChanged` boolean over updating immediately?

Reply:
I had just recent rebased with that changed around the same time as your comment.

The boolean usage is to prevent an update if `changedSpace` is already at the front of `recentSpaces`, I think it's the only use case where we don't have to update.

Reply:
It was also to separate the action of manipulating the list and saving the list; the flatMap returns the formatted list of spaces that can be saved when subscribed to afterwards. I can remove the boolean though and move the saving.

Reply:
Oh, that's valid; I just wanted to know the reason. Sorry for making it sound like I wanted it to be changed. It's up to you, whichever you think is better here works.

Reply:
It's not really the emission that I'm worried about, it's the patch request to update the profile from saveRecent that I want to eliminate. I haven't dug around enough to figure out exactly what happens, but it seems that if you send a certain number of requests in a limited amount of time, then the profile doesn't get updated correctly. i.e., the more spaces (or contexts) I visit in quick succession, the less likely the profile is able to properly be updated with the values to persist a refresh.

Reply:
Ah right, that makes sense. I'll try to help brainstorm "better" ways to do this than `recentChanged`, but that might end up being the simplest way to do it. I would just really prefer to avoid having any kind of side effect in the mapping, even if it's as simple as setting essentially a dirty flag.

Reply:
Right now it's chaining off of the broadcast's separately and into the recent, I think if it goes from the recent and internalizes reading broadcasts inside into emitting recents, it could be cleaner. Sorry for the confusing description. I haven't thought about how to actualize it in code.

Reply:
@andrewazores with regards to removing recentChanged, what if the ReplaySubject was switched to be a BehaviorSubject instead? This way we could use a filter at the beginning of the observable chain and exit if the condition isn't satisfied.

e.g.,
```
this.broadcaster.on<Space>('spaceChanged')
  .filter((changedSpace: Space) => {
    return this._recent.getValue()[0].id !== changedSpace.id;
  })
  .flatMap((changedSpace: Space) => { ...
```
and
```
this.broadcaster.on<Space>('spaceDeleted')
  .filter((deletedSpace: Space) => {
    return this._recent.getValue().findIndex((space: Space) => space.id === deletedSpace.id) !== -1;
  }) ...
...
```

The `flatMap` wouldn't need to toggle the boolean and the subscription at the end could simply be the call to `saveRecent()`.

Reply:
I had an idea thinking back on my `distinctUntilChanged` thought here again. Doesn't that still meet our needs here? ie remove `recentChanged` completely, and continue performing `saveRecent` within the subscriber, but simply insert a `distinctUntilChanged` to transform the `flatMap` result before the result is passed in to the subscriber. The way I had it written out (and thought about) before was assuming consumers might or might not care about duplicate emissions - so why don't we just treat our own subscriber as a consumer, and use provided operators to filter the duplicates?

Reply:
Ha, we were writing our comments at the same time. Yea, using a `filter` instead of `distinctUntilChanged` could work too. Whichever comes out cleaner, really.